apps/iir_blur/iir_blur_generator.cpp

```diff
@@ -63,33 +63,41 @@ Func blur_cols_transpose(Func input, Expr height, Expr alpha, bool skip_schedule
             // Really for an IIR on the GPU you should use a more
             // specialized DSL like RecFilter, but we can schedule it
             // in Halide adequately, we just can't extract any
-            // parallelism from the scan dimension.
-
-            // We also have no way to store kernel intermediates on
-            // the GPU in global memory, and we can't fuse together
-            // multiple update stages into a single loop, so we
-            // currently have to launch a whole lotta
-            // kernels. store_with would ameliorate this to some
-            // degree, or making store_in get you to global memory on
-            // the GPU. All in all, this is a pretty mediocre schedule
-            // for an IIR blur.
-
-            // 3.39ms on a 2060 RTX
+            // parallelism from the scan dimension. Most GPUs will be
+            // heavily under-utilized with this schedule and thus
+            // unable to hide the memory latencies to L2.
 
+            // 2.06ms on a 2060 RTX
             Var xi, yi;
-            blur.compute_root();
+            transpose.compute_root()
+                .tile(x, y, xi, yi, 32, 32)
+                .gpu_blocks(y, c)
+                .gpu_lanes(xi);
+
+            blur.compute_at(transpose, y)
+                .store_in(MemoryType::Heap)  // Too large to fit into shared memory
+                .gpu_lanes(x);
             blur.update(0)
-                .split(x, x, xi, 32)
-                .gpu_blocks(x, c)
-                .gpu_threads(xi);
+                .gpu_lanes(x);
+
+            // We can't hide load latencies by swapping in other warps
+            // because we don't have enough available parallelism for
+            // that, but if we unroll the scan loop a little then the
+            // ptx compiler can reorder the loads earlier than the
+            // fmas, and cover latency that way. Saves 1.7ms!
             blur.update(1)
-                .split(x, x, xi, 32)
-                .gpu_blocks(x, c)
-                .gpu_threads(xi);
+                .unroll(ry, 8)
+                .gpu_lanes(x);
             blur.update(2)
-                .split(x, x, xi, 32)
-                .gpu_blocks(x, c)
-                .gpu_threads(xi);
+                .unroll(ry, 8)
+                .gpu_lanes(x);
+
+            // Stage the transpose input through shared so that we do
+            // strided loads out of shared instead of global.
+            blur.in()
+                .align_storage(x, 33)  // Avoid bank conflicts. Saves 0.05 ms
```

Comment:
33 is a weird magic number here; would be worthy of naming and explaining?

Reply:
Done, it's the smallest integer greater than the warp size which is coprime with the warp size - isn't that obvious!? Honestly I'm not even sure bank conflicts are still a thing, but it did reduce runtime slightly.